### PR TITLE
Expose soft body pin methods to GDScript

### PR DESCRIPTION
**Pull request by bengtsts**
Makes it easy to auto pin or unpin soft body points using vertex colours with a simple script and other like tools.

### CODE REVIEW

File: scene/3d/soft_body_3d.cpp
Comment (lyuma) [inline scene/3d/soft_body_3d.cpp:364]: add an argument `"point_index"`

File: scene/3d/soft_body_3d.cpp
Comment (lyuma) [inline scene/3d/soft_body_3d.cpp:366]: Same: add an argument `"point_index"`

File: scene/3d/soft_body_3d.cpp
Comment (lyuma) [inline scene/3d/soft_body_3d.cpp:368]: Same: add an argument `"point_index"`

File: scene/3d/soft_body_3d.cpp
Comment (lyuma) [inline scene/3d/soft_body_3d.cpp:367]: Add arguments `"point_index", "pin", "spatial_attachment_path"`

**Review by lyuma — CHANGES_REQUESTED** [review]
this isn't a review of the proposal itself, just correcting a mistake I noticed with how the methods were bound.

**Comment by lyuma** [comment]
I'd support exposing these functions.

Just to be clear on why this PR is useful... while an API does exist to pin points already exists via property paths, for example given a PackedInt32Array pinned_points:
```gd
	var should_pin: Array[bool] = .......
	
	var mesh_arrays: Array = new_node.mesh.surface_get_arrays(0) # Godot SoftBody ignores other surfaces.
	var mesh_verts: PackedVector3Array = mesh_arrays[Mesh.ARRAY_VERTEX]
	var pinned_points: PackedInt32Array = PackedInt32Array()
	for vert_idx in range(len(mesh_verts):
		if should_pin[vert_idx]:
			pinned_points.push_back(vert_idx)
	new_node.set("pinned_points", pinned_points)
```
the problem with the above is that you can't toggle a pinned point off, without replacing the whole array, which is no good from a performance perspective (setting attachments/x/point_index to -1 is actually not implemented)... and for the OP's realtime vertex painting application, that wouldn't be feasible.

Side note, since it's not documented: currently, the only way to pin a mesh with a skin is significantly more complicated and an array of NodePaths to BoneAttachment's. Here is a simple example without the BoneAttachment part:
```gd
	var bones_paths: Array = [].duplicate()
	var offsets: Array = [].duplicate()
	for vert_idx in range(len(mesh_verts):
		if should_pin[vert_idx]:
			bones_paths.push_back(NodePath("."))
			offsets.push_back(mesh_verts[vert_idx])
			vert_idx += 1
	for i in range(len(pinned_points)):
		new_node.set("attachments/" + str(i) + "/spatial_attachment_path", bones_paths[i])
		new_node.set("attachments/" + str(i) + "/offset", offsets[i])
```
On any reasonably large mesh, running the above will likely lag the inspector to death once all the NodePath's are created.

So while this PR is an improvement, the current APIs do not scale to large meshes, and I think godot is in need of a proposal for a different SoftBody api that would avoid storing a NodePath for every vertex...

**Review by pouleyKetchoupp — CHANGES_REQUESTED** [review]
Sounds like a good idea! It's just missing documentation update here:
https://github.com/godotengine/godot/blob/master/doc/classes/SoftBody3D.xml

**Edit:** Static checks are also failing on code formatting, please check the log for details.

**Comment by pouleyKetchoupp** [comment]
Static checks are failing because of an extra tab in code formatting:
https://github.com/godotengine/godot/pull/51364/checks?check_run_id=3455503191

For info, you can use commit hooks to catch this kind of issue more easily in your local branch, check how to install them here:
https://github.com/godotengine/godot/tree/master/misc/hooks

File: scene/3d/soft_body_3d.cpp
Comment (pouleyKetchoupp) [inline scene/3d/soft_body_3d.cpp:366]: `pin_point_toggle` makes sense for the editor gizmo, but it seems redundant for scripts since there's already a function to set it to true or false, so I would rather remove this one.

File: scene/3d/soft_body_3d.cpp
Comment (pouleyKetchoupp) [inline scene/3d/soft_body_3d.cpp:367]: You can make the function names consistent for scripts and add the default value like this (so when you pinned is false you don't have to set the attachment path):
```suggestion
	ClassDB::bind_method(D_METHOD("set_point_pinned", "point_index", "pinned", "spatial_attachment_path", DEFVAL(NodePath())), &SoftBody3D::pin_point);
```

That would only rename the script function.
You can also rename the method in SoftBody3D class to clean things up, but it's not necessary for this PR.

File: doc/classes/SoftBody3D.xml
Comment (pouleyKetchoupp) [inline doc/classes/SoftBody3D.xml:95]: Documentation formatting:
```suggestion
				Returns [code]true[/code] if vertex is set to pinned.
```

File: doc/classes/SoftBody3D.xml
Comment (pouleyKetchoupp) [inline doc/classes/SoftBody3D.xml:88]: Suggestion to make the description of arguments a bit more explicit:
```suggestion
				Sets the pinned state of a surface vertex. When set to [code]true[/code], the optional [code]spatial_attachment_path[/code] can define a spatial node the pinned vertex will be attached to.
```

**Review by pouleyKetchoupp — CHANGES_REQUESTED** [review]

File: scene/3d/soft_body_3d.cpp
Comment (bengtsts) [reply scene/3d/soft_body_3d.cpp:367]: If I'm not mistake, this rename will also update the mono and GDNative APIs, keeping them all consistent?

File: scene/3d/soft_body_3d.cpp
Comment (bengtsts) [reply scene/3d/soft_body_3d.cpp:366]: I don't see any downside to having this exposed. If someone just wants to toggle some pins, theres no need to rewrite logic thats been in the engine all this time.

Comment (pouleyKetchoupp) [reply scene/3d/soft_body_3d.cpp:366]: Yeah, you don't need to remove the method entirely from the engine, it's just not necessary to expose it to scripts. It can be confusing to have redundant functions, and especially this one is not that useful in game script context.

Comment (bengtsts) [reply scene/3d/soft_body_3d.cpp:366]: Perhaps not so much for games, _at this time_,  but surely for addons. Godot's non-script tools for editing pinned points are quite limited. The pin click can be expanded to a sphere volume brush or selection box/lasso.

I don't understand how its confusing or redundant either. Without a toggle method, the logic to check, change, and set the pin flag would have to be made to recreate a toggle. And if reused multiple times, that logic would be turned into a method. That seems redundant to me.

In any case, if you still don't agree, I won't press the matter further.

Comment (pouleyKetchoupp) [reply scene/3d/soft_body_3d.cpp:366]: Given that the same thing can be done with `set_point_pinned(point_index, !is_point_pinned(point_index))` I can't see a good reason to bloat the script API with an extra function.

The only cases where a helper function would be justified is if it's not trivial, like some math that users would have to figure out otherwise, or something that requires many lines of script.

File: scene/3d/soft_body_3d.cpp
Comment (pouleyKetchoupp) [reply scene/3d/soft_body_3d.cpp:367]: Yep, it will also update other script APIs but I don't think there's anything you need to worry about.

File: doc/classes/SoftBody3D.xml
Comment (pouleyKetchoupp) [inline doc/classes/SoftBody3D.xml:79]:
```suggestion
			<argument index="2" name="attachment_path" type="NodePath" default="NodePath(&quot;&quot;)" />
```

**Review by pouleyKetchoupp — CHANGES_REQUESTED** [review]
Thanks! Apart from the little detail from my comment it should be good to go.

**Comment by bengtsts** [comment]
@pouleyKetchoupp Because of the SoftBody3D and Node3D renaming, should I make another PR for 3.x once this is approved?

**Comment by pouleyKetchoupp** [comment]
@bengtsts Yes please, that would be awesome!

**Comment by pouleyKetchoupp** [comment]
Also just for info, in general you create a specific branch (local and remote) for a PR instead of using your master branch, so that it makes it easier if you need to work on your branch or create other PRs before this one is merged.

You don't have to change anything for this PR since it's going to be merged soon, but just something to keep in mind for next time :)

**Comment by pouleyKetchoupp** [comment]
Documentation is failing because it expects alphabetical order, I had missed that.

**Review by pouleyKetchoupp — APPROVED** [review]
All good now. Thanks and congrats on your first contribution!

**Comment by pouleyKetchoupp** [comment]
@lyuma Merge seems to be blocked by requested changes even though it's all resolved. Please approve to see if it helps :)

**Review by lyuma — APPROVED** [review]

**Comment by lyuma** [comment]
My sincerest apologies: I had assumed that "bengtsts marked this conversation as resolved." would resolve the requested changes and did not intend to block review :-p

It looks great.

**Comment by pouleyKetchoupp** [comment]
Thanks!
I thought so too and it's the first time I see that happen, something probably went wrong with github :)

**Pull request merged**
